aten/src/ATen/native/cpu/group_norm_kernel.cpp

```diff
@@ -461,114 +558,285 @@ void ComputeInternalGradients(
+        const T_ACC gamma_v = gamma_null ? T_ACC(1) : T_ACC(gamma[g * D + j]);
         ds_val += ds_ptr[j] * gamma_v;
         db_val += db_ptr[j] * gamma_v;
       }
-      const T c2 =
-          (db_val * mean[i] - ds_val) * rstd[i] * rstd[i] * rstd[i] * s;
-      const T c3 = -c2 * mean[i] - db_val * rstd[i] * s;
+      const T_ACC c2 =
+          (db_val * T_ACC(mean[i]) - ds_val) * T_ACC(rstd[i]) * T_ACC(rstd[i]) * T_ACC(rstd[i]) * s;
+      const T_ACC c3 = -c2 * T_ACC(mean[i]) - db_val * T_ACC(rstd[i]) * s;
+
       for (const auto j : c10::irange(D)) {
         const int64_t c = g * D + j;
         const T* dY_ptr = dY + (i * D + j) * HxW;
         const T* X_ptr = X + (i * D + j) * HxW;
         T* dX_ptr = dX + (i * D + j) * HxW;
-        const T c1 = rstd[i] * (gamma_null ? T(1) : gamma[c]);
+        const T_ACC c1 = T_ACC(rstd[i]) * (gamma_null ? T_ACC(1) : T_ACC(gamma[c]));
         for (const auto k : c10::irange(HxW)) {
-          dX_ptr[k] = c1 * dY_ptr[k] + c2 * X_ptr[k] + c3;
+          dX_ptr[k] = c1 * T_ACC(dY_ptr[k]) + c2 * T_ACC(X_ptr[k]) + c3;
         }
       }
     }
   });
 }
 
-template <typename T>
+template <typename PT, typename T_ACC>
 void GammaBackward(
     int64_t N,
     int64_t C,
     int64_t group,
-    const T* mean,
-    const T* rstd,
-    const T* ds,
-    const T* db,
-    T* dgamma) {
+    const PT* mean,
+    const PT* rstd,
+    const T_ACC* ds,
+    const T_ACC* db,
+    PT* dgamma) {
   const int64_t G = group;
   const int64_t D = C / G;
-  constexpr int64_t K = vec::Vectorized<T>::size();
+  constexpr int64_t K = vec::Vectorized<PT>::size();
   at::parallel_for(0, D, K, [=](int64_t start, int64_t end) {
     for (const auto i : c10::irange(G)) {
-      std::memset(dgamma + i * D + start, 0, (end - start) * sizeof(T));
+      std::memset(dgamma + i * D + start, 0, (end - start) * sizeof(PT));
     }
     for (int64_t i = 0; i < N * G; ++i) {
-      const T* ds_ptr = ds + i * D;
-      const T* db_ptr = db + i * D;
+      const T_ACC* ds_ptr = ds + i * D;
+      const T_ACC* db_ptr = db + i * D;
       const int64_t g = i % G;
+
       for (const auto j : c10::irange(start, end)) {
         const int64_t c = g * D + j;
-        dgamma[c] += (ds_ptr[j] - db_ptr[j] * mean[i]) * rstd[i];
+        dgamma[c] += (ds_ptr[j] - db_ptr[j] * T_ACC(mean[i])) * T_ACC(rstd[i]);
       }
     }
   });
 }
 
-template <typename T>
-void BetaBackward(int64_t N, int64_t C, const T* db, T* dbeta) {
-  constexpr int64_t K = vec::Vectorized<T>::size();
+template <>
+void GammaBackward(
+    int64_t N,
+    int64_t C,
+    int64_t group,
+    const BFloat16* mean,
+    const BFloat16* rstd,
+    const float* ds,
+    const float* db,
+    BFloat16* dgamma) {
+  using fVec = vec::Vectorized<float>;
+  using bVec = vec::Vectorized<BFloat16>;
+  const int64_t G = group;
+  const int64_t D = C / G;
+  constexpr int64_t K = bVec::size();
+  float* acc_buffer = new float[C];
```

Comment:
better to use unique_ptr?

Reply:
Remove the specialized version for PT = BFloat16 since parameters should not be bfloat16 in training